### ISSUE TITLE
`ConfigurateTranslator.translateMapOrList` throws NPE

### ISSUE THREAD

**Issue by XakepSDK** [SpongePowered/Sponge#1963]
- SpongeVanilla version: spongevanilla-1.12.2-7.1.0-BETA-74
  
`ConfigurateTranslator.translateMapOrList` throws NPE if node doesn't contain `key`.
Here is check if `key` is `null`: 
https://github.com/SpongePowered/SpongeCommon/blob/bleeding/src/main/java/org/spongepowered/common/data/persistence/ConfigurateTranslator.java#L123-L124
But inside this method there is `node.getKey().XXX` calls.
https://github.com/SpongePowered/SpongeCommon/blob/bleeding/src/main/java/org/spongepowered/common/data/persistence/ConfigurateTranslator.java#L83

So it throws NPE for me:
```
java.lang.NullPointerException: null
	at org.spongepowered.common.data.persistence.ConfigurateTranslator.translateMapOrList(ConfigurateTranslator.java:83) ~[ConfigurateTranslator.class:1.12.2-7.1.0-BETA-74]
	at org.spongepowered.common.data.persistence.ConfigurateTranslator.addTo(ConfigurateTranslator.java:124) ~[ConfigurateTranslator.class:1.12.2-7.1.0-BETA-74]
	at org.spongepowered.common.data.persistence.ConfigurateTranslator.translateFromNode(ConfigurateTranslator.java:71) ~[ConfigurateTranslator.class:1.12.2-7.1.0-BETA-74]
	at org.spongepowered.common.data.persistence.ConfigurateTranslator.translate(ConfigurateTranslator.java:115) ~[ConfigurateTranslator.class:1.12.2-7.1.0-BETA-74]
	at org.spongepowered.common.data.persistence.ConfigurateTranslator.translate(ConfigurateTranslator.java:45) ~[ConfigurateTranslator.class:1.12.2-7.1.0-BETA-74]
	at org.spongepowered.common.config.DataSerializableTypeSerializer.deserialize(DataSerializableTypeSerializer.java:50) ~[DataSerializableTypeSerializer.class:1.12.2-7.1.0-BETA-74]
	at org.spongepowered.common.config.DataSerializableTypeSerializer.deserialize(DataSerializableTypeSerializer.java:41) ~[DataSerializableTypeSerializer.class:1.12.2-7.1.0-BETA-74]
```

@gabizou

**Comment by XakepSDK** [comment]
This happens with this deserializer:
https://gist.github.com/XakepSDK/a213106357e1a177cb6c4b450c6275b0
on key deserialization.
`Map.Entry<>` contains this:
```
[03:34:08 INFO] [STDOUT]: 
{
    ContentVersion=1, ItemType=minecraft:apple, Count=1, UnsafeDamage=0
}
=
SimpleCommentedConfigurationNode{
    super=SimpleConfigurationNode{
        key={
            ContentVersion=1, ItemType=minecraft:apple, Count=1, UnsafeDamage=0
        }, 
        value=ListConfigValue{
            values=[
                SimpleCommentedConfigurationNode{
                    super=SimpleConfigurationNode{
                        key=0, value=ScalarConfigValue{value=TYPE}
                    }, 
                    comment=null
               }
            ]
        }
    }, 
    comment=null
}
```

**Comment by XakepSDK** [comment]
Well, whole issue is that key is deserialized as String, not map.

**Comment by simon816** [comment]
I discovered this a while back https://forums.spongepowered.org/t/nullpointerexception-loading-configuration-after-initial-save/16648/6?u=simon816
Guess I didn't make a GH issue about it.

Minimal example:
```java
DataTranslators.CONFIGURATION_NODE.translate(SimpleConfigurationNode.root().setValue("test"));
```

**Comment by ryantheleach** [comment]
What is the expected result? (I mean, clearly not throwing a NPE)?

Should this instead be moved to Configurates issue tracker.

**Comment by dualspiral** [comment]
I have backed-out the commit that was meant to fix this due to data loss and other issues. I'm reopening this so it doesn't get lost.

**Comment by ImMorpheus** [comment]
Fixed by #2603

A test was also added for this to avoid any regression:

https://github.com/SpongePowered/SpongeCommon/blob/16b0cfa3572f9946a94a3ece56e56c14d4189503/src/test/java/org/spongepowered/common/data/translator/ConfigurateDataViewTest.java#L194
